src/java/org/apache/cassandra/utils/concurrent/ExtendedNonBlockingHashMap.java

```diff
@@ -0,0 +1,202 @@
+            }
+
+            try
+            {
+                future.syncUninterruptibly();
+            }
+            catch (Throwable t)
+            {
+                // if blockingUnloadIfPresent was called in the meantime, we simply retry hoping that unloading gets
+                // finished soon
+                // also we retry if the concurrent attempt to load entry failed (but we do not retry if this attempt
+                // failed)
+                if (newEntry == null || Throwables.isCausedBy(t, ex -> ex instanceof KeyNotFoundException))
+                {
+                    Thread.yield();
+                    continue;
+                }
+            }
+
+            future.rethrowIfFailed();
+            return future.getNow();
+        }
+    }
+
+    /**
+     * If a value for the given key is present, unload function is run and the value is removed from the map.
+     * Similarly to {@link #blockingLoadIfAbsent(Object, Supplier)} at-most-once semantics is guaranteed for unload
+     * function.
+     * <p>
+     * When unload function fails, the value is removed from the map anyway and the failure is rethrown.
+     * <p>
+     * When the key was not found, the method returns {@code null}.
+     *
+     * @throws UnloadExecutionException when the unloading failed to complete - this is checked exception because
+     *                                  the value is removed from the map regardless of the result of unloading; therefore if the unloading failed, the
+     *                                  called is responsible for handling that; the {@link UnloadExecutionException} encapsulates the value which was
+     *                                  failed to unload.
+     */
+    public V blockingUnloadIfPresent(K key, Consumer<? super V> unloadFunction) throws UnloadExecutionException
+    {
+        Promise<V> droppedFuture = new AsyncPromise<V>().setFailure(new KeyNotFoundException());
+
+        Future<V> existingFuture;
+        do
+        {
+            existingFuture = get(key);
+            if (existingFuture == null || existingFuture.cause() != null)
+                return null;
+        } while (!replace(key, existingFuture, droppedFuture));
+
+        V v = existingFuture.awaitUninterruptibly().getNow();
+        if (v == null)
+        {
+            // which means that either the value failed to load or a concurrent attempt to unload already did the work
+            return null;
+        }
+
+        try
+        {
+            unloadFunction.accept(v);
+            return v;
+        }
+        catch (Throwable t)
+        {
+            throw new UnloadExecutionException(v, t);
+        }
+        finally
+        {
+            Future<V> future = remove(key);
+            assert future == droppedFuture;
+        }
+    }
+
+    private static class KeyNotFoundException extends IllegalStateException
```

Comment:
I would name this `ConcurrentUnloadException`. Since you are catching and processing this in the loading method, though, it should not be necessary -- a completed `null` future would work just as well with slightly simpler code.

Reply:
replacing with null value sounds like a great idea